cpp/src/internal/request_am.cpp

```diff
@@ -0,0 +1,38 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include <ucxx/buffer.h>
+#include <ucxx/delayed_submission.h>
+#include <ucxx/internal/request_am.h>
+#include <ucxx/request_am.h>
+
+namespace ucxx {
+
+namespace internal {
+
+RecvAmMessage::RecvAmMessage(internal::AmData* amData,
+                             ucp_ep_h ep,
+                             std::shared_ptr<RequestAm> request,
+                             std::shared_ptr<Buffer> buffer)
+  : _amData(amData), _ep(ep), _request(request), _buffer(buffer)
+{
+  _request->_delayedSubmission =
+    std::make_shared<DelayedSubmission>(false, _buffer->data(), _buffer->getSize());
+}
+
+void RecvAmMessage::setUcpRequest(void* request) { _request->_request = request; }
+
+void RecvAmMessage::callback(void* request, ucs_status_t status)
+{
+  _request->_buffer = _buffer;
```

Comment:
Can `_request` be used for something else during the lifetime of `this`? if not, why not set up the `_request->_buffer` pointer in the constructor of `RecvAmMessage`, if yes, then this seems potentially racy.

Reply:
That's a good idea, done in https://github.com/rapidsai/ucxx/pull/48/commits/6eb5f95b48c9f407d32a9a29267082ecaf1884bd.